src/Core/Types/FunctionType.cs

```diff
@@ -90,6 +90,25 @@ public override DataType Clone()
             return ft;
 		}
 
+        public bool IsVarargs()
+        {
+            var last = Parameters != null ? Parameters.LastOrDefault() : null;
+            return last != null && last.Name == "...";
+        }
+
+        public FunctionType ReplaceVarargs(params Identifier[] parameters)
+        {
+            if (!IsVarargs())
+                throw new NotSupportedException(
+                    "Signature should contain varargs");
+            var sig = (FunctionType)Clone();
+            sig.Parameters = Parameters.
+                Where(a => a.Name != "...").
+                Union(parameters).
+                ToArray();
```

Comment:
Since parameter lists are ordered collections, you're better off using the `.Concat()` extension method rather than `.Union()`. The latter is fully justified in converting `(a, a).Union(a, a)` into `(a)` which isn't correct.